pkg/metrics/collector/collector.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2018 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This package is modified from prometheus node_exporter
+
+package collector
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var procPath string
+
+var (
+	scrapeDurationDesc = prometheus.NewDesc(
+		"scrape_collector_duration_seconds",
+		"Duration of a collector scrape.",
+		[]string{"collector"},
+		nil,
+	)
+	scrapeSuccessDesc = prometheus.NewDesc(
+		"scrape_collector_success",
+		"Whether a collector succeeded.",
+		[]string{"collector"},
+		nil,
+	)
+	// Needed by prometheus-to-sd
+	scrapeStartDesc = prometheus.NewDesc(
+		"process_start_time_seconds",
+		"Start time of the process since unix epoch in seconds.",
+		nil,
+		nil,
+	)
+	startTime int64
+)
+
+var (
+	factories = make(map[string]func() (Collector, error))
+)
+
+func registerCollector(collector string, factory func() (Collector, error)) {
+	factories[collector] = factory
+}
+
+// NodeCollector implements the prometheus.Collector interface.
+type nodeCollector struct {
+	Collectors map[string]Collector
+}
+
+func init() {
+	startTime = time.Now().Unix()
+}
+
+// NewNodeCollector creates a new NodeCollector
+func NewNodeCollector(filters []string, proc string) (*nodeCollector, error) {
+	procPath = proc
+	collectors := make(map[string]Collector)
+	for _, filter := range filters {
+		createFunc, exist := factories[filter]
+		if !exist {
+			return nil, fmt.Errorf("missing collector: %s", filter)
+		} else {
+			if c, err := createFunc(); err == nil {
+				collectors[filter] = c
+			} else {
+				return nil, fmt.Errorf("fail to create collector: %s, error %v", filter, err)
```

Comment:
it's better to return err directly, than wrap it
```go
c, err := createFunc()
if err != nil {
  return nil, err
}
collectors[filter] = c
```

Reply:
Updated